### PR TITLE
[aws-datstore][DRAFT] Prototype of user-provided configs

### PR DESCRIPTION
**Pull request by jamesonwilliams**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: aws-datastore/src/main/java/com/amplifyframework/datastore/DiscardingConflictHandler.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    @Override
+    public <T extends Model> void resolveConflict(
+            @NonNull DataStoreConflictData<T> conflictData,
+            @NonNull Consumer<DataStoreConflictHandlerResult> onConflictResolved) {
+        final T local = conflictData.getLocal();
+        final T remote = conflictData.getRemote();
+        Amplify.DataStore.delete(local,
+            deleted -> Amplify.DataStore.save(remote,
+                saved -> onConflictResolved.accept(DataStoreConflictHandlerResult.DISCARD),
+                dataStoreErrorHandler
```
Comment (jamesonwilliams) [inline aws-datastore/src/main/java/com/amplifyframework/datastore/DiscardingConflictHandler.java:58]: If this all fails, then `onConflictResolved` is never called back. Perhaps we need `DataStoreConflictHandlerResult.FAILURE`?

File: aws-datastore/src/test/java/com/amplifyframework/datastore/AWSDataStorePluginTest.java [hunk 2]
```diff
@@ -66,6 +67,7 @@ public void setup() {
      * @throws JSONException Not expected; on failure to arrange configuration object
      * @throws DataStoreException Not expected; on failure to configure of initialize plugin
      */
+    @Ignore("TODO: need a mechanism to re-enable this scenario........")
```
Comment (jamesonwilliams) [inline aws-datastore/src/test/java/com/amplifyframework/datastore/AWSDataStorePluginTest.java:70]: .e.g., we need to inspect state of API category to understand if it has a suitable configuration for our use .... (what are the appropriate checks for this, that don't leak abstraction? In reality, it's gonna break hard if it is not an `AWSApiPlugin`. But in fantasy land,  we want a flat dependency tree where-in plugins _do not_ take a dependency on one another.)

File: aws-datastore/src/main/java/com/amplifyframework/datastore/AWSDataStorePlugin.java [hunk 3]
```diff
@@ -132,18 +129,10 @@ public String getPluginKey() {
...
-            throw new DataStoreException(
-                "There was an issue configuring the plugin from the amplifyconfiguration.json",
-                badConfigException,
-                "Check the attached exception for more details and " +
-                    "be sure you are only calling Amplify.configure once"
-            );
-        }
+    ) {
+        // AWSDataStorePlugin does not consider any values from amplifyconfiguration.json.
+        // To effect the behavior of the plugin, provide an {@link DataStoreConfiguration}
```
Comment (wooj2) [inline aws-datastore/src/main/java/com/amplifyframework/datastore/AWSDataStorePlugin.java:134]: Hmm i'm pretty sure we should: s/effect/affect

Also, this is a bit surprising to me... but, have we decided we are not going to accept values from amplifyconfiguration.json ?

File: aws-datastore/src/main/java/com/amplifyframework/datastore/AWSDataStorePlugin.java [hunk 4]
```diff
@@ -441,4 +423,127 @@ private void afterInitialization(@NonNull final Runnable runnable) {
...
+             * Configures the {@link ModelSchemaRegistry} into which schema will be kept at runtime.
+             * @param modelSchemaRegistry The registry into which {@link ModelSchema} will be stored
+             * @return The next step in the build process
+             */
+            GraphQlBehaviorStep modelSchemaRegistry(@NonNull ModelSchemaRegistry modelSchemaRegistry);
+        }
+
+        /**
+         * The step where an {@link GraphQlBehavior} is specified. After this, all components are
+         * specified, and the building precedes to offer the {@link BuildStep#build()} as a final
```
Comment (wooj2) [inline aws-datastore/src/main/java/com/amplifyframework/datastore/AWSDataStorePlugin.java:513]: preceds, or proceeds?

File: aws-datastore/src/main/java/com/amplifyframework/datastore/AWSDataStorePlugin.java [hunk 5]
```diff
@@ -441,4 +423,127 @@ private void afterInitialization(@NonNull final Runnable runnable) {
...
+             * Configures the {@link GraphQlBehavior} that is used to talk to the AppSync endpoint.
+             * This component will only be used if remote model synchronization is enabled in the plugin.
+             * @param graphQlBehavior A GraphQL client which can be used to talk to AppSync.
+             * @return The step of the Builder where an AWSDataStorePlugin finally gets constructed
+             */
+            ConfigurationStep graphQlBehavior(@NonNull GraphQlBehavior graphQlBehavior);
+        }
+
+        /**
+         * The step of the building process where the user provides a configuration for the DataStore.
```
Comment (wooj2) [inline aws-datastore/src/main/java/com/amplifyframework/datastore/AWSDataStorePlugin.java:527]: s/for the Datastore/for Datastore/

File: aws-datastore/src/main/java/com/amplifyframework/datastore/DataStoreConfiguration.java [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+         * @return Current builder
+         */
+        @NonNull
+        public Builder dataStoreConflictHandler(@NonNull DataStoreConflictHandler dataStoreConflictHandler) {
+            this.dataStoreConflictHandler = Objects.requireNonNull(dataStoreConflictHandler);
+            return Builder.this;
+        }
+
+        /**
+         * Sets a handler function to be applied when the DataStore encounters an unrecoverable error
```
Comment (wooj2) [inline aws-datastore/src/main/java/com/amplifyframework/datastore/DataStoreConfiguration.java:150]: s/the DataStore/DataStore

File: aws-datastore/src/main/java/com/amplifyframework/datastore/DataStoreConfiguration.java [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+         * @return Current builder instance
+         */
+        @NonNull
+        public Builder syncIntervalMs(@IntRange(from = 0) long syncIntervalMs) {
+            this.syncIntervalMs = syncIntervalMs;
+            return Builder.this;
+        }
+
+        /**
+         * Sets the maximum number of records, from the server, to process from a sync operation.
```
Comment (wooj2) [inline aws-datastore/src/main/java/com/amplifyframework/datastore/DataStoreConfiguration.java:173]: might want to point out that this is number of records per model

File: aws-datastore/src/main/java/com/amplifyframework/datastore/DataStoreConflictHandler.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+package com.amplifyframework.datastore;
+
+import androidx.annotation.NonNull;
+
+import com.amplifyframework.core.Consumer;
+import com.amplifyframework.core.model.Model;
+
+/**
+ * Handles conflicts between models observed during model synchronization in the DataStore.
```
Comment (wooj2) [inline aws-datastore/src/main/java/com/amplifyframework/datastore/DataStoreConflictHandler.java:24]: Hmm this is the third time i've seen this.. is it really "the Datastore" ?  Maybe it's intentional?

File: aws-datastore/src/main/java/com/amplifyframework/datastore/DataStoreConflictHandler.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+ * Such conflicts may arise between version of models kept locally, and competing versions
+ * found in the remote AppSync system.
```
Comment (wooj2) [inline aws-datastore/src/main/java/com/amplifyframework/datastore/DataStoreConflictHandler.java:26]: s/system/service

File: aws-datastore/src/main/java/com/amplifyframework/datastore/DataStoreConfiguration.java [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+     * a base/delta sync operation from AppSync.
+     * @return The max number of records to process from AppSync.
+     */
+    @IntRange(from = 0)
+    public int getSyncMaxRecords() {
+        return this.syncMaxRecords;
+    }
+
+    /**
+     * Gets the number of items that should be requested in page, from AppSync, during
```
Comment (wooj2) [inline aws-datastore/src/main/java/com/amplifyframework/datastore/DataStoreConfiguration.java:109]: "in a page"

File: aws-datastore/src/main/java/com/amplifyframework/datastore/AWSDataStorePlugin.java [hunk 11]
```diff
@@ -81,10 +80,19 @@ private AWSDataStorePlugin(
...
 
+    /**
+     * Begin building a new {@link AWSDataStorePlugin} instance, by means of a new
+     * {@link Builder} instance.
+     * @return The first step in a sequence of builder actions.
+     */
+    public static BuilderSteps.ModelProviderStep builder() {
+        return new Builder();
+    }
+
```
[the remark this answers is not in the document]
Comment (jamesonwilliams) [reply aws-datastore/src/main/java/com/amplifyframework/datastore/AWSDataStorePlugin.java:95]: (I redacted several paragraphs in favor of an external link)

TLDR: Lombok provides utility to reduce boiler plate in Java, but there are better alternatives, including [Autovalue](https://github.com/google/auto/blob/master/value/userguide/builders.md), [Immutables](https://immutables.github.io/), or even [Kotlin data clases](https://kotlinlang.org/docs/reference/data-classes.html). (Ref: ["Don't use Lombok."](https://medium.com/@vgonzalo/dont-use-lombok-672418daa819))

File: aws-datastore/src/main/java/com/amplifyframework/datastore/DataStoreConfiguration.java (hunk 7 above)
Comment (jamesonwilliams) [reply aws-datastore/src/main/java/com/amplifyframework/datastore/DataStoreConfiguration.java:173]: Ah! Right. That was not even clear in my mind, as a wrote this. But is, in retrospect, true.

File: aws-datastore/src/main/java/com/amplifyframework/datastore/DataStoreConflictHandler.java (hunk 8 above)
Comment (jamesonwilliams) [reply aws-datastore/src/main/java/com/amplifyframework/datastore/DataStoreConflictHandler.java:24]: Months back, we had decided on the stylization _DataStore_ as a product decision.

So, we laid to reset the Pascal-casing identifier for the product. However, I do not think we have standardized on the camel-cased and sreaming-snake sytlizations:

0. `dataStore`, or `datastore`?
1. `DATA_STORE` or `DATASTORE`?

IIRC, Android is ~uniformly using these stylizations, right now:
- `DataStore` (pascal)
- `dataStore` (camel), and
- `DATASTORE` (screaming snake.)

**Comment by jamesonwilliams** [comment]
Closing this in favor of @rjuliano's more evolved https://github.com/aws-amplify/amplify-android/pull/395. @rjuliano -- @wooj2 caught a bunch of documentation nits in his review, can we make sure those are included into your more comprehensive  https://github.com/aws-amplify/amplify-android/pull/395?
